src/test/ui/deref-suggestion.rs

```diff
@@ -0,0 +1,29 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+/*fn foo(_: String) {}
+
+fn foo2(s: &String) {
+    foo(s);
+}*/
+
+fn foo3(_: u32) {}
+fn foo4(u: &u32) {
+    foo3(u);
+}
+
+fn main() {
+    let s = String::new();
+    let r_s = &s;
+    foo(r_s);
+    foo(&"aaa".to_owned());
+    foo(&mut "aaa".to_owned());
+    foo4(&0);
```

Comment:
I saw some logic around macros -- can we embed this in a macro and see what happens?

Reply:
e.g. `foo(borrow!(0))` where:

```
macro_rules! borrow {
   ($x:expr) => { &$x }
}
```

Reply:
Good idea!